cpp/src/io/csv/csv_reader.cu

```diff
@@ -207,6 +212,80 @@ std::string stringType(gdf_dtype dt){
 
 }
 
+/**
+ * reads contents of an arrow::io::RandomAccessFile in a char * buffer up to the number of bytes specified in bytes_to_read
+ * for non local filesystems where latency and availability can be an issue it will retry until it has exhausted its the read attemps and empty reads that are allowed
+ */
+gdf_error read_file_into_buffer(std::shared_ptr<arrow::io::RandomAccessFile> file, int64_t bytes_to_read, uint8_t* buffer, int total_read_attempts_allowed, int empty_reads_allowed){
+
+	if (bytes_to_read > 0){
+
+		int64_t total_read;
+		arrow::Status status = file->Read(bytes_to_read,&total_read, buffer);
+
+		if (!status.ok()){
+			return GDF_FILE_ERROR;
+		}
+
+		if (total_read < bytes_to_read){
+			//the following two variables shoudl be explained
+			//Certain file systems can timeout like hdfs or nfs,
+			//so we shoudl introduce the capacity to retry
+			int total_read_attempts = 0;
+			int empty_reads = 0;
+
+			while (total_read < bytes_to_read && total_read_attempts < total_read_attempts_allowed && empty_reads < empty_reads_allowed){
+				int64_t bytes_read;
+				status = file->Read(bytes_to_read-total_read,&bytes_read, buffer + total_read);
+				if (!status.ok()){
+					return GDF_FILE_ERROR;
+				}
+				if (bytes_read == 0){
+					empty_reads++;
+				}
+				total_read += bytes_read;
+			}
+			if (total_read < bytes_to_read){
+				return GDF_FILE_ERROR;
+			} else {
```

Comment:
Why do you need three `else {  return GDF_SUCCESS; }`? Just fall through to a single return success.